### PR TITLE
Moves random source generation into helper library

### LINKED ISSUE #7159: Flaky stress test in VB suite.
The following test in the VB suite fails erratically with an `OutOfMemoryException` on `builder.ToString()`.

Didn't we already fix this in the C# suite? Why didn't we do it here as well?

```vb
    <WorkItem(2867, "https://github.com/dotnet/roslyn/issues/2867")>
    <Fact>
    Public Sub TestBinary()
        ' use a fixed seed so the test Is reproducible
        Dim random = New System.Random(12345)
        Const n As Integer = 40 * 1000 * 1000 ' 40 million "character"s
        Dim builder = New System.Text.StringBuilder(n + 10)
        For i As Integer = 0 To n - 1
            builder.Append(ChrW(random.Next(&HFFFF)))
        Next

        Dim source = builder.ToString()
        Dim tree = VisualBasicSyntaxTree.ParseText(source)

        Assert.Equal(source, tree.ToString())
        Assert.Equal(Syntax.InternalSyntax.Scanner.BadTokenCountLimit, tree.GetDiagnostics().Where(Function(d) d.Code = ERRID.ERR_IllegalChar).Count())
    End Sub
```

### PR DESCRIPTION
**Pull request by TyOverby**
Closes #7159

### CODE REVIEW

**Comment by TyOverby** [comment]
@gafter @dotnet/roslyn-compiler

File: src/Compilers/VisualBasic/Test/Syntax/BasicCompilerSyntaxTest.vbproj [hunk 1]
```diff
@@ -96,7 +96,7 @@
     <Compile Include="IncrementalParser\SyntaxDifferences.vb" />
     <Compile Include="LocationTests.vb" />
     <Compile Include="Parser\DeclarationTests.vb" />
-    <Compile Include="Parser\FuzzTesting.vb" />
+    <Compile Include="Parser\ParserRegressionTests.vb" />
```
Comment (jmarolf) [inline src/Compilers/VisualBasic/Test/Syntax/BasicCompilerSyntaxTest.vbproj]: :+1:

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+
+namespace Roslyn.Test.Utilities.Syntax
+{
+    public static class SourceUtilities
+    {
+        public static String GenerateLargeRandomString()
```
Comment (jaredpar) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: Use `string` instead of `String`

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
```
Comment (jaredpar) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: Need a copyright header

**Comment by davkean** [comment]
Linux failure due to #7411, I've taken the machine offline.

@dotnet-bot retest prtest/lin/dbg/unit32 please

**Comment by AlekseyTs** [comment]
LGTM

**Comment by jaredpar** [comment]
:+1:

**Comment by TyOverby** [comment]
@gafter wants me to remove this test entirely (along with the corresponding C# one).  Apparently his machine fails the C# version of this test (with the fix!) one out of ten times, making it a huge hinderance to getting work done.

**Comment by jaredpar** [comment]
We should focus on making this valuable test more reliable vs. deleting it.  

Why is this test failing on his machine.  Is it an OOM?  If so I'm guessing it's because it requires a contiguous allocation.  Have we considered breaking it up into chunks, deriving from `SourceText` and driving the parser that way?

**Comment by TyOverby** [comment]
@jaredpar Yeah, it's failing on OOM. 

> deriving from SourceText

That's a good idea.  I'll play with an implementation.

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs (hunk 2 above)
Comment (TyOverby) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: Fixed

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs (hunk 3 above)
Comment (TyOverby) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: Done

**Comment by TyOverby** [comment]
@jaredpar @gafter:

This should fix the flakiness.  

I did what Jared suggested and subclassed SourceText to be a list of smaller buffers.

**Comment by TyOverby** [comment]
@dotnet/roslyn-compiler 

I made a pretty big change since the last reviews.  If I could get someone to look over this code again, that would be great!

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            var i = 0;
+
+            foreach (char ch in from)
+            {
+                if (i == BLOCK_SIZE)
+                {
+                    i = 0;
+                    blocks.Add(new char[BLOCK_SIZE]);
+                }
+                blocks.Last()[i] = ch;
```
Comment (AlekseyTs) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: Consider capturing in a local the current block that we are filling instead of calling blocks.Last() for each character.

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+        public override Encoding Encoding => Encoding.UTF8;
+
+        public override int Length
+        {
+            get
+            {
+                if (blocks.Count != 0)
+                {
+                    return (blocks.Count - 1) * BLOCK_SIZE + blocks.Last().Length;
```
Comment (AlekseyTs) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: Is there an expectation that ```blocks.Last().Length``` could be different from BLOCK_SIZE?

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            }
+        }
+
+        public override char this[int position]
+        {
+            get
+            {
+                var block = position / BLOCK_SIZE;
+                var offset = position % BLOCK_SIZE;
+                return blocks[block][offset];
```
Comment (AlekseyTs) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: It looks like an assumption is made that the last block is completely filled with data? What this assumption is based on?

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+            for(int i = 0; i < n; i++)
+            {
+                yield return (char)random.Next(char.MaxValue);
+            }
+        }
+
+        public static SourceText GenerateLargeRandomString()
+        {
+            return new SegmentedSourceText(RandomString());
```
Comment (AlekseyTs) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: It looks like SegmentedSourceText keeps all the characters in memory, can this be avoided? For example, if the tests that are using this SourceText never access characters in random order, but always do this consecutively, we could have implementation of SourceText that doesn't cache any characters in memory, but rather "generates" them on demand.

Comment (TyOverby) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: I'll look into it.

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+using System.Text;
+
+namespace Roslyn.Test.Utilities.Syntax
+{
+    internal sealed class RandomizedSourceText: SourceText
+    {
+        private char hash(int position)
+        {
+            // good enough "random looking" hasher via the xorshift128plus algorithm.
+            UInt64 state0 = (UInt64) 1;
```
Comment (gafter) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: This does not appear to be related to the data previously processed by the test.

**Comment by TyOverby** [comment]
@AlekseyTs: This PR now streams random characters into the lexer and parser.  No more allocations at all!

File: src/Compilers/VisualBasic/Test/Syntax/Parser/ParserRegressionTests.vb [hunk 9]
```diff
@@ -918,20 +919,8 @@ End Enum
...
-        Dim builder = New System.Text.StringBuilder(n + 10)
-        For i As Integer = 0 To n - 1
-            builder.Append(ChrW(random.Next(&HFFFF)))
-        Next
-
-        Dim source = builder.ToString()
-        Dim tree = VisualBasicSyntaxTree.ParseText(source)
-
-        Assert.Equal(source, tree.ToString())
+        Dim tree = VisualBasicSyntaxTree.ParseText(New RandomizedSourceText())
```
Comment (gafter) [inline src/Compilers/VisualBasic/Test/Syntax/Parser/ParserRegressionTests.vb]: Do we have some reason to believe that this modified test would have triggered the same (original) issue?

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs [hunk 10]
```diff
@@ -83,24 +81,7 @@ public void c024928()
...
-                {
-                    for(int i = 0; i < n; i++)
-                    {
-                        chars[i] = (char)random.Next(char.MaxValue);
-                    }
-                }
-            }
-
-            var tree = CSharpSyntaxTree.ParseText(str);
+            CSharpSyntaxTree.ParseText(new RandomizedSourceText());
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs]: Do we have some reason to believe that this modified test would have triggered the same (original) issue?

The (flakiness) failure I frequently see in this test is in the debug code inside the parser written to verify that the parsed text is consistent with the original; it creates a 40MB string to do that, which exhausts memory when tests are run in parallel. This doesn't really change that.

File: src/Compilers/VisualBasic/Test/Syntax/Parser/ParserRegressionTests.vb (hunk 9 above)
Comment (gafter) [inline src/Compilers/VisualBasic/Test/Syntax/Parser/ParserRegressionTests.vb]: OK, I'm willing to believe it would.

Comment (TyOverby) [inline src/Compilers/VisualBasic/Test/Syntax/Parser/ParserRegressionTests.vb]: The original issue was "we try to read binary files, and we generate millions of errors to report back to the user; this crashes us."  

Although this doesn't use the same seed, it will trigger the original issue.

**Comment by TyOverby** [comment]
test this please

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs (hunk 10 above)
Comment (TyOverby) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs]: This code gets to the parsing phase?  I thought for sure that the errors that we generate come before we have a parse tree to validate.

File: src/Compilers/VisualBasic/Test/Syntax/Parser/ParserRegressionTests.vb (hunk 9 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Syntax/Parser/ParserRegressionTests.vb]: This sounds reasonable to me.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs (hunk 10 above)
Comment (gafter) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs]: Most of the failures I've seen recently have been in the parser where it validates that the input == the syntax tree stringified.

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs (hunk 8 above)
Comment (TyOverby) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: The specific random data from the previous test was mostly irrelevant.  The only thing that mattered was generating more syntax errors than we could handle.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs (hunk 10 above)
Comment (TyOverby) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs]: Really?  How do we get a parse tree if the text is literally unparsable?  What parse tree would be generated by the text "if if if if if if if if \0"?

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs [hunk 11]
```diff
@@ -83,24 +81,7 @@ public void c024928()
         [Fact]
         public void TestBinary()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs]: Note, there is another PR that disables the test make sure the test is enabled once this change is merged.

File: src/Compilers/VisualBasic/Test/Syntax/Parser/ParserRegressionTests.vb [hunk 12]
```diff
@@ -918,20 +919,8 @@ End Enum
     End Sub
 
     <WorkItem(2867, "https://github.com/dotnet/roslyn/issues/2867")>
-    <Fact(Skip:="https://github.com/dotnet/roslyn/issues/7159")>
     Public Sub TestBinary()
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Syntax/Parser/ParserRegressionTests.vb]: The ```Fact``` attribute must be preserved without the Skip reason.

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+            UInt64 state1 = (UInt64) position;
+            unchecked
+            {
+                var s1 = state0;
+                var s0 = state1;
+                state0 = s0;
```
Comment (AlekseyTs) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: It looks like ```state0``` is not used after this point. Is there a reason to change its value? Can it be a const local?

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+                s1 ^= s1 << 23;
+                s1 ^= s1 >> 17;
+                s1 ^= s0;
+                s1 ^= s0 >> 26;
+                state1 = s1;
```
Comment (AlekseyTs) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: Is there a particular reason why we need ```state1``` local? It looks like we can simply initialize ```s0``` with ```position``` and then use ```s1``` in place of ```state1``` in the return statement. If I am not missing something, the code will be clearer.

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+                return (char)(state1 + s0);
+            }
+        }
+
+        public override char this[int position] => hash(position);
```
Comment (AlekseyTs) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: Should we throw on index out of range?

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+            {
+                return 40 * 1000 * 1000;
+            }
+        }
+
+        public override void CopyTo(int sourceIndex, char[] destination, int destinationIndex, int count)
+        {
+            for (var i = 0; i < count; i++)
+            {
+                destination[destinationIndex + i] = hash(sourceIndex);
```
Comment (AlekseyTs) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: It looks like we fill buffer with the same character, ```sourceIndex``` is never incremented. Also, should we throw throw on an attempt to get characters out of range?

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs (hunk 11 above)
Comment (TyOverby) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs]: Yep, I'll pay attention during merge.  Fortunately Git should flag it as a conflict.

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs (hunk 15 above)
Comment (TyOverby) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: We could?  It would slow things down and not provide any extra test coverage though.

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs (hunk 16 above)
Comment (TyOverby) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: > `sourceIndex` is never incremented

Fixed, thanks.

> Also, should we throw throw on an attempt to get characters out of range?

I doubt this would improve test coverage of that case, but adding it would be easy enough.

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs (hunk 14 above)
Comment (TyOverby) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: The "reason" is that I naively copied pseudocode from a wikipedia article.

I've cleaned up the locals and also changed state0 and state1 to be better chosen "constants".

File: src/Compilers/VisualBasic/Test/Syntax/Parser/ParserRegressionTests.vb (hunk 12 above)
Comment (TyOverby) [inline src/Compilers/VisualBasic/Test/Syntax/Parser/ParserRegressionTests.vb]: Done.

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs (hunk 14 above)
Comment (cston) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: Can we avoid this approach and do something simpler? Perhaps ```return builder[position % builder.Count()]``` where `builder` is 1K version of the buffer we were creating originally.

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs (hunk 16 above)
Comment (gafter) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: Can you please push your changes into the PR?

Comment (TyOverby) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: Yep; sorry.  Got busy and forgot about this.

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+        {
+            get
+            {
+                return 40 * 1000 * 1000;
+            }
+        }
+
+        public override void CopyTo(int sourceIndex, char[] destination, int destinationIndex, int count)
+        {
+            throw new UnauthorizedAccessException();
```
Comment (AlekseyTs) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: Is implementation of this method not needed for the test? Then we should probably throw NotImplementedException or NotSupportedException and get rid of the remaining code.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs (hunk 10 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs]: We are producing parse trees for any text given as an input. Even when there are parsing/scaning errors, the parse tree is supposed produce the original text when converted to string.

File: src/Test/Utilities/Shared/Syntax/SourceUtilities.cs (hunk 17 above)
Comment (TyOverby) [inline src/Test/Utilities/Shared/Syntax/SourceUtilities.cs]: Oh, I threw that in there to see if `CopyTo` was being used.  Turns out it was (as you can see in the test failures).

When @gafter asked me to push my fixes to other issues, I forgot to change this back to it's actual implementation.

**Comment by TyOverby** [comment]
@gafter @AlekseyTs 

It seems like the only remaining problem is the

> debug code inside the parser written to verify that the parsed text is consistent with the original; it creates two 40MB strings to do that, one for the input and one for the parse tree. -- @gafter 

Do you think that we could have these tests run only on release builds?  This would keep testing for this scenario, but wouldn't do that crazy allocation.  (in fact, this would better test the real world use-case that we originally created this test for)

**Comment by AlekseyTs** [comment]
I think running on release only is a reasonable trade-off.

**Comment by TyOverby** [comment]
@dotnet/roslyn-infrastructure 

Where are all of these commits coming from?  Did someone rewrite history or did I mess things up?

**Comment by TyOverby** [comment]
run prtest/win/dbg/eta please

**Comment by TyOverby** [comment]
test prtest/win/dbg/eta please

File: src/Test/Utilities/Shared/Assert/ConditionalFactAttribute.cs [hunk 18]
```diff
@@ -49,4 +49,15 @@ public class IsEnglishLocal : ExecutionCondition
 
         public override string SkipReason => "Current culture is not en-US";
     }
+
+    public class IsRelease : ExecutionCondition
+    {
+#if DEBUG
+        public override bool ShouldSkip => true;
+#else
+        public override bool ShouldSkip => true;
```
Comment (AlekseyTs) [inline src/Test/Utilities/Shared/Assert/ConditionalFactAttribute.cs]: Should this return true?

Comment (AlekseyTs) [inline src/Test/Utilities/Shared/Assert/ConditionalFactAttribute.cs]: I mean false?

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs [hunk 19]
```diff
@@ -80,27 +78,10 @@ public void c024928()
         }
 
         [WorkItem(2771, "https://github.com/dotnet/roslyn/issues/2771")]
-        [Fact]
+        [ConditionalFact(typeof(IsRelease))]
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs]: Wouldn't it be simpler to surround this test with ```#if !DEBUG```?

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs [hunk 20]
```diff
@@ -79,28 +77,13 @@ public void c024928()
             var tree = SyntaxFactory.ParseSyntaxTree(test);
         }
 
+#if !DEBUG
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs]: Rather than use a conditional directive here, can we introduce `FactReleaseOnly`?  A custom `Fact` type is generally how we control tests that execute under specific conditions.

Comment (TyOverby) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs]: Ack! that's what I had and Aleksy recommended that I remove it in favor of this.  I can revert and fix the mistake that I had in that version.

Comment (TyOverby) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs]: Also, we must have "release-only" tests and "debug-only" tests, but I couldn't find any with a quick `git grep`.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs]: I suggested to use conditional compilation because the custom fact would also be implemented using conditional compilation, but in a different project. It felt natural to me to simply \#if the unit-test rather than build whole infrastructure to accomplish the same thing.

Comment (jaredpar) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs]: My reasoning for wanting to use a custom `Fact` is the following:

1. In general I dislike `#if` directives and want to minimize them because it makes it harder to accurately analyze source code as it creates multiple views of the code.  This is a frequent problem in tools like `codeformatter`.  If we run this tool with `Release` defined (which is a valid idea) then it will never process any of this added code.  
2. Our pattern thus far has been to conditionally execute tests by using a custom `Fact` attribute.  Using `#if` manually here goes against that pattern.

**Comment by gafter** [comment]
LGTM, though I expect you'll want to address @jaredpar 's comment about using a custom `Fact` attribute for release-only.

**Comment by TyOverby** [comment]
Alright, I moved back to using a custom Fact attribute.  If all these tests pass, I'll rewrite the commit message (I don't know how this happened!) and merge.

**Comment by TyOverby** [comment]
@jaredpar Could I get one more pass over this before I merge?

**Comment by jaredpar** [comment]
:+1:
